src/util.ts

```diff
@@ -0,0 +1,106 @@
+import * as sourcegraph from 'sourcegraph'
+import { Metadata } from './link-preview-expander'
+import LRU from 'lru-cache'
+
+export function getWord(document: sourcegraph.TextDocument, range: sourcegraph.Range): string {
+    const lines = document.text?.split('\n')
+    let word = ''
+    if (lines && range) {
+        word = lines[range.start.line].slice(range.start.character, range.end.character)
+    }
+    return word
+}
+
+export function checkIsURL(maybeURL: string): boolean {
+    try {
+        const url = new URL(maybeURL)
+        return url.protocol === 'http:' || url.protocol === 'https:'
+    } catch {
+        return false
+    }
+}
+
+/**
+ * Strips trailing comma or period from URL
+ *
+ * @param maybeURL Possible URL string
+ */
+export function cleanURL(maybeURL: string): string {
+    const lastCharacter = maybeURL.slice(-1)
+    if (lastCharacter === '.' || lastCharacter === ',') {
+        return maybeURL.slice(0, -1)
+    }
+    return maybeURL
+}
+
+export const FAILURE = Symbol('FAILURE')
+
+interface MetadataCache {
+    /** Report a failed fetch for a URL */
+    reportFailure: (URL: string) => void
+
+    /**
+     * Returns undefined when cached value is not of Metadata type and
+     * number of failures is below the retry threshold.
+     */
+    get: (URL: string) => Metadata | typeof FAILURE | undefined
+
+    /**
+     * Store successful metadata for a URL
+     */
+    set: (URL: string, metadata: Metadata) => void
+}
+
+/**
+ * Creates a simple LRU cache for metadata.
+ *
+ * Motivation:
+ * - It's easy for users to trigger multiple hovers unintentionally
+ * - Users may hover over the same URL several times to re-read previews
+ * - Don't want to blow through cors-anywhere rate-limit
+ * - Hover to presented metadata is a relatively expensive process.
+ * (fetch -> parse html -> merge content from relevant meta tags)
```

Comment:
Does the cache solve a concrete problem, or is this perhaps a bit premature optimization? Is there a reason the browser/fetch/HTTP cache does not suffice (which is configurable in the fetch options and persistent across page reloads)?

The caching code all feels a bit overengineered given that it's all gone after a page reload anyway. If we do need a cache I'd just make it a simple variable without any cache eviction.  Given we want these to be code examples of extensions I would try to keep them as simple as possible.

Reply:
Considering that these will be examples, I'll just use a plain object. 

My primary concern was the `cors-anywhere` rate-limit. Since only one token is selected, it's easy to trigger multiple requests at a time. However, once we have our own `cors-anywhere` instance set up, this won't matter.